kinesis/src/main/scala/akka/stream/alpakka/kinesis/KinesisSourceStage.scala

```diff
@@ -0,0 +1,164 @@
+            completeStage()
+          } { iterator = _ }
+          if (records.size > 0) {
+            records.foreach(buffer.enqueue(_))
+            self.become(ready)
+            self.ref ! Pump
+          } else {
+            scheduleOnce('GET_RECORDS, refreshInterval)
+          }
+        }
+        case ((_, GetRecordsFailure(ex))) => {
+          log.error("Failed to fetch records from Kinesis for shard {}", shardId)
+          failStage(Errors.GetRecordsError)
+        }
+        case ((_, Pump)) => ()
+      }
+
+      private def ready(in: (ActorRef, Any)): Unit = in match {
+        case ((_, Pump)) => {
+          if (isAvailable(shape.out)) {
+            push(shape.out, buffer.dequeue())
+            self.ref ! Pump
+          }
+          if (buffer.isEmpty) {
+            self.become(awaitingRecords)
+            requestRecords()
+          }
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case 'GET_RECORDS => requestRecords()
+      }
+
+      private[this] val onGetShardIteratorFailure = getAsyncCallback(
+        (ex: Throwable) => self.ref ! GetShardIteratorFailure(ex)
+      )
+
+      private[this] val onGetShardIteratorSuccess = getAsyncCallback(
+        (result: GetShardIteratorResult) => self.ref ! GetShardIteratorSuccess(result)
+      )
+
+      private[this] val getShardIteratorAsyncHandler =
+        getAmazonKinesisHandler[GetShardIteratorRequest, GetShardIteratorResult](onGetShardIteratorSuccess,
+                                                                                 onGetShardIteratorFailure)
+
+      private[this] val getRecordsFailureCallback = getAsyncCallback(
+        (ex: Throwable) => self.ref ! GetRecordsFailure(ex)
+      )
+
+      private[this] val getRecordsSuccessCallback = getAsyncCallback(
+        (result: GetRecordsResult) => self.ref ! GetRecordsSuccess(result)
+      )
+
+      private[this] val getRecordsAsyncHandler = getAmazonKinesisHandler[GetRecordsRequest, GetRecordsResult](
+        getRecordsSuccessCallback,
+        getRecordsFailureCallback
+      )
+
+      private[this] def requestRecords(): Unit =
+        amazonKinesisAsync.getRecordsAsync(new GetRecordsRequest().withShardIterator(iterator), getRecordsAsyncHandler)
+
+      private[this] def getShardIterator(): Future[GetShardIteratorResult] = {
+        val request = Function.chain(
+          Seq(
+            (r: GetShardIteratorRequest) => r.withStreamName(streamName),
+            (r: GetShardIteratorRequest) => r.withShardId(shardId),
+            (r: GetShardIteratorRequest) => r.withShardIteratorType(shardIteratorType),
+            (r: GetShardIteratorRequest) => startingSequenceNumber.fold(r)(r.withStartingSequenceNumber),
+            (r: GetShardIteratorRequest) => atTimestamp.fold(r)(r.withTimestamp)
+          )
+        )(new GetShardIteratorRequest())
```

Comment:
Nice usage of `Function.chain`. You can reduce verbosity by giving a type parameter to `chain` like so:

```scala
val request = Function.chain[GetShardIteratorRequest](
  Seq(
    _.withStreamName(streamName),
    ...
    r => atTimestamp.fold(r)(r.withTimestamp)
  )
)(new GetShardIteratorRequest())
```